packages/flutter_web_plugins/lib/src/plugin_registry.dart

```diff
@@ -0,0 +1,124 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:ui' as ui;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+
+typedef _MessageHandler = Future<ByteData> Function(ByteData);
+
+/// This class registers web platform plugins.
+class PluginRegistry {
+  /// Creates a plugin registry.
+  PluginRegistry(this._binaryMessenger);
+
+  final BinaryMessenger _binaryMessenger;
+
+  /// Creates a registrar for the given plugin implementation class.
+  Registrar registrarFor(Type key) => Registrar(_binaryMessenger);
+
+  /// Registers this plugin handler with the engine, so that unrecognized
+  /// platform messages are forwarded to the registry, where they can be
+  /// correctly dispatched to one of the registered plugins.
+  void registerMessageHandler() {
+    // The function below is only defined in the Web dart:ui.
+    // ignore: undefined_function
+    ui.webOnlySetPluginHandler(_binaryMessenger.handlePlatformMessage);
```

Comment:
I'm not sure how this will handle being rolled into google3. I got away with the network_image case by ignoring that file in the BUILD config

Reply:
I'm not sure how the google3 setup is, but the way I see it working is that this package will only be pulled in for Flutter web application builds. So, for plugins, there will be two targets, one `dart_library` for the client library, and one for the web-implementation (which will depend on `package:flutter_web_plugins`).